File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/Nonce.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bcrand.fillRandomData
import com.blockchaincommons.bctags.TAG_NONCE
import com.blockchaincommons.bcur.UR
import com.blockchaincommons.bcur.URCodable
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborTaggedCodable
import com.blockchaincommons.dcbor.CborTaggedUtils
import com.blockchaincommons.dcbor.Tag
import com.blockchaincommons.dcbor.tagsForValues

/**
 * A random nonce ("number used once").
 *
 * A [Nonce] is a 12-byte random value used in cryptographic protocols to
 * prevent replay attacks and ensure the uniqueness of encrypted messages.
 */
class Nonce private constructor(private val data: ByteArray) :
    CborTaggedCodable,
    URCodable {

    init {
        require(data.size == NONCE_SIZE) {
            "Nonce data must be exactly $NONCE_SIZE bytes, got ${data.size}"
        }
    }

    /** Returns a copy of the underlying nonce data. */
    fun data(): ByteArray = data.copyOf()

    /** Returns the nonce bytes as a copy. */
    fun asBytes(): ByteArray = data.copyOf()

    /** The nonce as a lowercase hexadecimal string. */
    val hex: String get() = data.toHexString()

    /** Returns the nonce bytes as a new [ByteArray]. */
    fun toByteArray(): ByteArray = data.copyOf()

    // -- equals / hashCode --

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Nonce) return false
        return data.contentEquals(other.data)
    }

    override fun hashCode(): Int = data.contentHashCode()

    // -- toString --

    override fun toString(): String = "Nonce($hex)"

    // -- CBOR --

    override fun cborTags(): List<Tag> = tagsForValues(listOf(TAG_NONCE))

    override fun untaggedCbor(): Cbor = Cbor.fromByteString(data)

    // -- Companion --

    companion object {
        const val NONCE_SIZE: Int = 12

        /** Creates a new random nonce using the secure random number generator. */
        fun create(): Nonce {
            val data = ByteArray(NONCE_SIZE)
            fillRandomData(data)
            return Nonce(data)
        }

        /** Restores a nonce from exactly [NONCE_SIZE] bytes. */
        fun fromData(data: ByteArray): Nonce {
            require(data.size == NONCE_SIZE) {
                "Nonce data must be exactly $NONCE_SIZE bytes, got ${data.size}"
            }
            return Nonce(data.copyOf())
        }

        /**
         * Restores a nonce from a byte array, throwing if the length is wrong.
         */
        fun fromDataChecked(data: ByteArray): Nonce {
            if (data.size != NONCE_SIZE) {
                throw BcComponentsException.invalidSize("nonce", NONCE_SIZE, data.size)
            }
            return Nonce(data.copyOf())
        }

        /**
         * Creates a nonce from a hexadecimal string.
         *
         * @throws IllegalArgumentException if the string is not exactly 24
         *   hex digits.
         */
        fun fromHex(hex: String): Nonce = fromData(hex.hexToByteArray())

        /** Decodes a [Nonce] from untagged CBOR (a byte string). */
        fun fromUntaggedCbor(cbor: Cbor): Nonce {
            val bytes = cbor.tryByteStringData()
            return fromDataChecked(bytes)
        }

        /** Decodes a [Nonce] from tagged CBOR. */
        fun fromTaggedCbor(cbor: Cbor): Nonce =
            CborTaggedUtils.fromTaggedCbor(
                cbor,
                tagsForValues(listOf(TAG_NONCE)),
            ) { fromUntaggedCbor(it) }

        /** Decodes a [Nonce] from tagged CBOR binary data. */
        fun fromTaggedCborData(data: ByteArray): Nonce =
            CborTaggedUtils.fromTaggedCborData(
                data,
                tagsForValues(listOf(TAG_NONCE)),
            ) { fromUntaggedCbor(it) }

        /** Decodes a [Nonce] from a UR. */
        fun fromUr(ur: UR): Nonce {
            ur.checkType("nonce")
            return fromUntaggedCbor(ur.cbor)
        }

        /** Decodes a [Nonce] from a UR string. */
        fun fromUrString(urString: String): Nonce =
            fromUr(UR.fromUrString(urString))
    }
}

File: csharp/BCComponents/BCComponents/Nonce.cs
using BlockchainCommons.BCRand;
using BlockchainCommons.BCTags;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// A random nonce ("number used once").
/// </summary>
/// <remarks>
/// A <see cref="Nonce"/> is a cryptographic primitive consisting of a random
/// or pseudo-random number that is used only once in a cryptographic
/// communication. Nonces are often used in authentication protocols, encryption
/// algorithms, and digital signatures to prevent replay attacks and ensure
/// the uniqueness of encrypted messages.
///
/// In this implementation, a nonce is a 12-byte random value. The size is
/// chosen to be sufficiently large to prevent collisions while remaining
/// efficient for storage and transmission.
/// </remarks>
public sealed class Nonce : IEquatable<Nonce>, ICborTaggedEncodable, ICborTaggedDecodable
{
    /// <summary>The size of a nonce in bytes.</summary>
    public const int Size = 12;

    private readonly byte[] _data;

    private Nonce(byte[] data)
    {
        _data = data;
    }

    /// <summary>
    /// Generates a new random nonce using the system's cryptographically secure RNG.
    /// </summary>
    /// <returns>A new random <see cref="Nonce"/>.</returns>
    public static Nonce New()
    {
        var data = new byte[Size];
        SecureRandomNumberGenerator.SecureFillRandomData(data);
        return new Nonce(data);
    }

    /// <summary>
    /// Creates a nonce from a 12-byte array.
    /// </summary>
    /// <param name="data">Exactly 12 bytes of nonce data.</param>
    /// <returns>A new <see cref="Nonce"/>.</returns>
    /// <exception cref="BCComponentsException">Thrown if <paramref name="data"/> is not exactly 12 bytes.</exception>
    public static Nonce FromData(byte[] data)
    {
        if (data.Length != Size)
            throw BCComponentsException.InvalidSize("nonce", Size, data.Length);
        var copy = new byte[Size];
        Array.Copy(data, copy, Size);
        return new Nonce(copy);
    }

    /// <summary>Gets a copy of the nonce data as a 12-byte array.</summary>
    public byte[] Data => (byte[])_data.Clone();

    /// <summary>Returns the nonce data as a read-only span.</summary>
    /// <returns>A read-only span over the nonce bytes.</returns>
    public ReadOnlySpan<byte> AsBytes() => _data;

    /// <summary>
    /// Creates a nonce from a hexadecimal string.
    /// </summary>
    /// <param name="hex">A 24-character hexadecimal string.</param>
    /// <returns>A new <see cref="Nonce"/>.</returns>
    /// <exception cref="FormatException">Thrown if the hex string is invalid.</exception>
    /// <exception cref="BCComponentsException">Thrown if the decoded data is not exactly 12 bytes.</exception>
    public static Nonce FromHex(string hex)
    {
        var data = Convert.FromHexString(hex);
        return FromData(data);
    }

    /// <summary>Gets the nonce data as a lowercase hexadecimal string.</summary>
    public string Hex => Convert.ToHexString(_data).ToLowerInvariant();

    // --- ICborTagged ---

    /// <summary>Returns the CBOR tags associated with the Nonce type.</summary>
    public static IReadOnlyList<Tag> CborTags => GlobalTags.TagsForValues(BcTags.TagNonce);

    // --- ICborTaggedEncodable ---

    /// <summary>Returns the untagged CBOR representation (a byte string).</summary>
    public Cbor UntaggedCbor() => Cbor.ToByteString(_data);

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor TaggedCbor() => Cbor.ToTaggedValue(CborTags[0], UntaggedCbor());

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor ToCbor() => TaggedCbor();

    // --- ICborTaggedDecodable ---

    /// <summary>
    /// Decodes a <see cref="Nonce"/> from an untagged CBOR byte string.
    /// </summary>
    /// <param name="cbor">The untagged CBOR value.</param>
    /// <returns>A new <see cref="Nonce"/>.</returns>
    /// <exception cref="CborException">Thrown if the CBOR is not a valid byte string.</exception>
    /// <exception cref="BCComponentsException">Thrown if the data is not exactly 12 bytes.</exception>
    public static Nonce FromUntaggedCbor(Cbor cbor)
    {
        var data = cbor.TryIntoByteString();
        return FromData(data);
    }

    /// <summary>
    /// Decodes a <see cref="Nonce"/> from a tagged CBOR value.
    /// </summary>
    /// <param name="cbor">The tagged CBOR value.</param>
    /// <returns>A new <see cref="Nonce"/>.</returns>
    /// <exception cref="CborException">Thrown if the CBOR tag does not match or the data is invalid.</exception>
    public static Nonce FromTaggedCbor(Cbor cbor)
    {
        foreach (var tag in CborTags)
        {
            try
            {
                var item = cbor.TryIntoExpectedTaggedValue(tag);
                return FromUntaggedCbor(item);
            }
            catch (CborWrongTagException) { }
            catch (CborWrongTypeException) { }
        }
        throw new CborWrongTypeException();
    }

    // --- IEquatable<Nonce> ---

    /// <inheritdoc/>
    public bool Equals(Nonce? other)
    {
        if (other is null) return false;
        return _data.AsSpan().SequenceEqual(other._data);
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is Nonce n && Equals(n);

    /// <inheritdoc/>
    public override int GetHashCode()
    {
        var hash = new HashCode();
        foreach (var b in _data)
            hash.Add(b);
        return hash.ToHashCode();
    }

    /// <summary>Tests equality of two Nonce instances.</summary>
    public static bool operator ==(Nonce? left, Nonce? right) =>
        left is null ? right is null : left.Equals(right);

    /// <summary>Tests inequality of two Nonce instances.</summary>
    public static bool operator !=(Nonce? left, Nonce? right) => !(left == right);

    // --- Display ---

    /// <inheritdoc/>
    public override string ToString() => $"Nonce({Hex})";

    /// <summary>Returns a byte array copy of this nonce's data.</summary>
    /// <returns>A 12-byte array.</returns>
    public byte[] ToByteArray() => Data;
}
